"Pan-cancer inference of intra-tumor heterogeneity reveals associations with different forms of genomic instability", F. Raynaud, M. Mina, D. Tavernari and G. Ciriello
====================================================================================================================

These files are necessary to generate the supplementary table containing:

#Sample_name	#Cancer_type	#Cancer_subtype		#Mean_reads_per_mutations	#Number_of_mutations	#Number_of_altered_segments	#Number_of_clones	#TreeScore	#Number_of_mutations_first_clone		#Number_of_mutations_other_clones



REQUIREMENTS
===========
Phylogenies generated by PhyloWGS (PhyloWGS: Reconstructing subclonal composition and evolution from whole-genome sequencing of tumors](http://genomebiology.com/2015/16/1/35), Deshwar et al.)

Output files from PhyloWGS for each sample:

* top_k_trees which contains the best phylogenies (50 by default) with the label of each clone, the population frequency of each clone, the number of children of each clone, number of mutations in the clone, the labels of the mutations

* top_k_trees1, top_k_trees2, ... ,  top_k_treesN
output file for the best k trees (50 by default)


FILES
=====
*Molecular data for the tumor types: CESC, UCEC, UVM, THCA, KICH, BRCA, SKCM, ACC, CRC, STAD, BLCA, LUAD, KIRP, PRAD, LIHC; has been collected in July 2015
*Molecular data for the tumor types: SARC, PAAD, MESO, LGG, GBM, DLBC, UCS, THYM, TGCT, PCPG, OV, LUSC, LAML, KIRC, HNSC, ESCA, CHOL; has been collected in 2018

from the FireHose (https://gdac.broadinstitute.org/) and cBioPortal (Cerami et al., 2012) (http://www.cbioportal.org/) data repositories for The Cancer Genome Atlas (TCGA). Only TCGA datasets publicly available at that time were used in our study. 

-Mutation files (MAF format): combined_2015_2018_MAF.maf.bz2
-Copy number segmentation files: combined_seg_2015_2018.seg.bz2


-analyze_public.py: Python file to generate the Supplementary Table
run: python2.7 analyze_public.py
